Dashboard tab revision

qnkhuat opened the pull request:
Epic: https://github.com/metabase/metabase/issues/29502
This PR implements revision to support dashboard tabs


<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30898)
<!-- Reviewable:end -->

qnkhuat committed:
Tab revision

qnkhuat committed:
typo

qnkhuat committed:
fix tests

calherries commented on src/metabase/api/dashboard.clj:
```diff
@@ -654,6 +621,47 @@
    [:id   ms/Int]
    [:name ms/NonBlankString]])
 
+(defn- do-cards-and-tabs-updates-tracking!
```
Why did you have to do this refactor to split out the tracking from the updates? It's not immediately obvious to me

qnkhuat replied:
I thought it'll be nice to refactor this because this PR adds 3 new events type, which is half of how many events "could" be fired on a single update.

and I think going forward we should follow this pattern, to me it has 2 advantages:
- the tracking is done after the transaction succeed
- readability, it's quite hard to trace what will be tracked in the current form IMO.


but if you find it's out of scope of this PR then I can do the refactor in a separated PR.

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -536,41 +537,15 @@
+                                                                       (map #(assoc % :dashboard_id dashboard-id))))]
     (zipmap (map :id new-tabs) new-tab-ids)))
 
-(mu/defn ^:private update-tabs! :- nil?
```
moved to `metabase.models.dashboard-tab`

calherries commented on test/metabase/api/revision_test.clj:
```diff
@@ -142,37 +142,37 @@
-               :diff                 nil
-               :has_multiple_changes false
-               :description          "created this."}]
+      (is (=? [{:is_reversion         true
```
This change makes me slightly confused what exactly should be included in the here. Either we care about all the keys under `:diff`, or we don't care about any. I think we should have all the keys here, including `dashboard_tab_id` to be safe

qnkhuat replied:
removed =? and fill in the missing fields in 0cbbda7595

calherries commented on src/metabase/models/dashboard.clj:
```diff
@@ -183,36 +183,62 @@
+                               (assoc :series (mapv :id (dashboard-card/series dashboard-card)))))))
+      (assoc :tabs (map #(select-keys % [:id :name :position :dashboard_id]) (ordered-tabs dashboard)))))
+
+(defn- revert-tabs
```
The bulk of this function looks suspiciously like `metabase.api.dashboard/do-update-tabs!`. I think you can do a refactor to this function so it takes a dashboard-id instead of a dashboard, and move it to the `metabase.models.dashboard-tab` namespace. Then you can do something like this:

```clojure 
(defn- revert-tabs
  [dashboard-id serialized-tabs]
  (let [current-tabs             (t2/select-fn-vec #(dissoc :created_at :updated_at :entity_id) :model/DashboardTab
                                                   :dashboard_id dashboard-id)
        {:keys [old-id->new-id]} (dashboard-tab/do-update-tabs! dashboard-id current-tabs serialized-tabs)]
    {:old-id->new-id old-id->new-id}))
```

qnkhuat replied:
moved it to `do-updated-tabs` and reused both in API and revert

calherries commented on src/metabase/models/dashboard.clj:
```diff
@@ -183,36 +183,62 @@
+         (dashboard-tab/update-tabs! (map id->current-tab update-ids) (map id->serialized-tab update-ids)))
+   {:old-id->new-id  (zipmap (map :id to-create-tabs) new-tab-ids)}))
+
+(defn- revert-dashcards
```
Also realised this looks very similar to `metabase.api.dashboard/do-update-dashcards!`. There's opportunity for reuse here.

qnkhuat replied:
initially I wanted to do that, but the `create-dashcards!` or `update-dashcards` need to do [param checking](https://github.com/metabase/metabase/blob/b64cd2674581716b5ce68186fee62c2c46f47b83/src/metabase/api/dashboard.clj#L603) that calls to too many fns in the API ns that I can't move it out.

calherries commented on test/metabase/models/dashboard_test.clj:
```diff
@@ -29,168 +30,204 @@
+                      {:id 2 :name "Third tab" :position 2}]}
+              {:tabs [{:id 0 :name "First tab" :position 0}]}))))
 
+   (is (= "rearranged the tabs."
```
You need to add test cases for "modified the tabs" - both for renaming the tabs and adding and removing them at the same time

qnkhuat replied:
added 0cbbda7595

qnkhuat committed:
Resolves Cal's suggestions

qnkhuat commented on src/metabase/models/dashboard_tab.clj:
```diff
@@ -102,3 +113,48 @@
     (t2/delete! :model/DashboardTab :id [:in tab-ids]))
   nil)
+
+(mu/defn classify-changes :- [:map
```
not sure where to put this tbh, we can put it in `metabase.util` but it's currently written in `cljc` and it doesn't support malli schema :/

calherries replied:
Can't you put it in a reader macro inside `metabase.util`?

qnkhuat commented on src/metabase/models/dashboard.clj:
```diff
@@ -183,36 +183,62 @@
-                               (assoc :series (mapv :id (dashboard-card/series dashboard-card)))))))))
+                           (-> (select-keys dashboard-card [:dashboard_tab_id :size_x :size_y :row :col :id :card_id])
+                               (assoc :series (mapv :id (dashboard-card/series dashboard-card)))))))
+      (assoc :tabs (map #(select-keys % [:id :name :position :dashboard_id]) (ordered-tabs dashboard)))))
```
yep, updated here 0cbbda7595

qnkhuat commented on src/metabase/models/dashboard.clj:
```diff
@@ -183,36 +183,62 @@
+(defn- revert-dashcards
+  [dashboard-id user-id serialized-cards]
+  (let [id->serialized-card (zipmap (map :id serialized-cards) serialized-cards)
+        current-cards       (t2/select [DashboardCard :id :size_x :size_y :row :col :card_id :dashboard_id]
+                                       :dashboard_id dashboard-id)
```
I agree, but lemme fix that all together in https://github.com/metabase/metabase/pull/31118

qnkhuat commented on src/metabase/models/dashboard.clj:
```diff
@@ -183,36 +183,62 @@
+
+(defn- revert-tabs
+  [dashboard-id serialized-tabs]
+  (let [current-tabs                       (t2/select [:model/DashboardTab :id :name :position]
```
updated in 0cbbda7595

qnkhuat commented on src/metabase/models/dashboard_tab.clj:
```diff
@@ -59,19 +59,46 @@
+  [current-tabs :- [:sequential [:map [:id ms/PositiveInt]]]
+   new-tabs     :- [:sequential [:map [:id ms/PositiveInt]]]]
+  (let [update-ks       [:name :position]
+        id->current-tab (group-by :id current-tabs)
```
good catch, updated in updated in 0cbbda7595

qnkhuat commented on test/metabase/events/revision_test.clj:
```diff
@@ -168,13 +172,90 @@
+      (is (= {:model        "Dashboard"
+              :model_id     dashboard-id
+              :user_id      (mt/user->id :rasta)
+              :object       (assoc (dashboard->revision-object dashboard) :tabs [])
```
removed

qnkhuat commented:
@calherries 

> If you create a dashboard, add a tab, then save, the revision history will say "You created 2 tabs", even though you just created one.

I don't think that's a bug, because you actually just created 2 tabs: one initial tab + the tab you just created.

maybe it's sound counter-intuitive but I think the proper fix is to have all new dashboard come with a tab.

calherries commented on src/metabase/models/dashboard_tab.clj:
```diff
@@ -59,19 +61,100 @@
+(defn do-update-tabs!
+  "Given current tabs and new tabs, do the necessary create/update/delete to apply new tab changes.
+  Returns:
+  - a map of temporary tab ID to the new real tab ID
+  - a list of deleted tab ids"
```
This list needs expanding and should include the keys in the result map:
```suggestion
  - `old->new-tab-id`: a map from tab IDs in `new-tabs` to newly created tab IDs
```

calherries commented on src/metabase/models/dashboard_tab.clj:
```diff
@@ -59,19 +61,100 @@
+    (when (seq to-update)
+      (update-tabs! current-tabs to-update))
+    {:old->new-tab-id old->new-tab-id
+     :created-tab-ids (map :id to-create)
```
This seems incorrect since `to-create` contains the old ids, not the new ones.
```suggestion
     :created-tab-ids (vals old->new-tab-id)
```

calherries commented on src/metabase/models/dashboard_tab.clj:
```diff
@@ -59,19 +61,100 @@
+     :to-delete (when (seq delete-ids) (filter #(delete-ids (:id %)) current-items))
+     :to-update (when (seq update-ids) (filter #(update-ids (:id %)) new-items))}))
+
+(defn do-update-tabs!
```
Found a bug. Imagine you start with no tabs, then add some cards to the dash. Save. This is version A. Then you create a tab, save, then delete it, save. So now you have one tab with some cards. Then revert to version A. The one remaining tab should get deleted, but the dashcard’s dashboard_tab_id ID field won’t update, and will continue referencing the deleted tab.

qnkhuat replied:
I'm not sure I'm following, can you loom it?

but I don't think it's possible for dashcard to reference to deleted tabs since we have an FK for it.

calherries replied:
Ah of course, you're right. Then the weird thing is the dashcards on the first tab will be deleted and will have to be created again, along with any associated series they have. This is probably OK as long as the dashcards are updated after the tabs are deleted. Not sure if it's worth complicating the code to fix this though, because it works and can't see how it could break.

calherries commented on src/metabase/api/dashboard.clj:
```diff
@@ -683,40 +635,32 @@
+                                           (seq deleted-tab-ids)
+                                           (remove (fn [card]
+                                                     (contains? deleted-tab-ids (:dashboard_tab_id card)))))
+                new-cards                (cond->> cards
+                                           ;; fixup the temporary tab ids with the real ones
+                                           (seq old->new-tab-id)
+                                           (map (fn [card]
+                                                  (if-let [real-tab-id (get old->new-tab-id (:dashboard_tab_id card))]
+                                                    (assoc card :dashboard_tab_id real-tab-id)
+                                                    card))))
```
Found a bug. Imagine you start with no tabs, then add some cards to the dash. Save. This is version A. Then you create a tab, save, then delete it, save. So now you have one tab with some cards. Then revert to version A. The one remaining tab should get deleted, but the dashcard’s dashboard_tab_id ID field won’t update, and will continue referencing the deleted tab. You'll need to also fixup dashboard_tab_id for tab IDs that have been deleted.

qnkhuat committed:
Cal's feedback

calherries commented on src/metabase/api/dashboard.clj:
```diff
@@ -683,40 +635,32 @@
                       {:status-code 400})))
     (api/check-500
-      (let [tabs-stats-atom (atom nil)]
+      (let [update-stats (atom nil)]
```
I think this name could be improved because it starts with `update-` which implies it is a function, not a noun. How about `changes-stats` or `changes-info`? `changes` has the benefit of not being specific to update operations.
```suggestion
      (let [changes-stats (atom nil)]
```

qnkhuat replied:
`changes` SGTM

qnkhuat replied:
updated in c7a6e58f97

calherries commented on src/metabase/api/dashboard.clj:
```diff
@@ -683,40 +635,32 @@
+                                                  (if-let [real-tab-id (get old->new-tab-id (:dashboard_tab_id card))]
+                                                    (assoc card :dashboard_tab_id real-tab-id)
+                                                    card))))
+                update-dashcards-stats   (do-update-dashcards! dashboard current-cards new-cards)]
```
Same as above
```suggestion
                dashcard-changes-stats   (do-update-dashcards! dashboard current-cards new-cards)]
```

calherries commented on src/metabase/api/dashboard.clj:
```diff
@@ -683,40 +635,32 @@
+                    (merge
+                      (select-keys update-tabs-stats [:created-tab-ids :updated-tab-ids :deleted-tab-ids :total-num-tabs])
+                      (select-keys update-dashcards-stats [:created-dashcards :deleted-dashcards :updated-dashcards])))))
+        (do-cards-and-tabs-updates-tracking! id @update-stats)
```
This could be more specific
```suggestion
        (track-dashcard-and-tab-events! id @changes-stats)
```

calherries commented on test/metabase/models/dashboard_test.clj:
```diff
@@ -29,168 +30,212 @@
+          (build-sentence
+            (revision/diff-strings
+              Dashboard
+              {:tabs [{:id 0 :name "First tab" :position 0}
+                      {:id 1 :name "Second tab" :position 1}]}
+              {:tabs [{:id 1 :name "First tab" :position 0}
+                      {:id 0 :name "Second tab" :position 1}]}))))
```
This test has a bug, and I believe the code does too. The only thing that should change is the position.
```suggestion
              {:tabs [{:id 0 :name "First tab" :position 0}
                      {:id 1 :name "Second tab" :position 1}]}
              {:tabs [{:id 1 :name "Second tab" :position 0}
                      {:id 0 :name "First tab" :position 1}]}))))
```

qnkhuat replied:
you're right, fixed by wrapping a `set` when checking name changes c7a6e58f97

calherries commented on test/metabase/models/dashboard_test.clj:
```diff
@@ -29,168 +30,212 @@
+          (build-sentence
+            (revision/diff-strings
+              Dashboard
+              {:tabs [{:id 0 :name "First tab" :position 0}
+                      {:id 1 :name "Second tab" :position 1}]}
+              {:tabs [{:id 1 :name "First tab new name" :position 0}
+                      {:id 0 :name "Second tab new name" :position 1}]}))))))
```
The point of this test is the tab IDs changed position and the name changed, but it's not super clear. This might make it clearer:
```suggestion
              {:tabs [{:id 0 :name "tab A" :position 0}
                      {:id 1 :name "tab B" :position 1}]}
              {:tabs [{:id 1 :name "tab B new name and position" :position 0}
                      {:id 0 :name "tab A new name and position" :position 1}]}))))))
```

qnkhuat replied:
updated in c7a6e58f97

calherries commented on test/metabase/models/dashboard_test.clj:
```diff
@@ -243,20 +291,229 @@
+
+     ;; revert
+     (revert-to-previous-revision Dashboard dashboard-id 2)
+     (testing "should works"
```
This just adds noise to the test description imo
```suggestion
```

qnkhuat replied:
removed  in c7a6e58f97

qnkhuat committed:
Cal's feedback

qnkhuat committed:
simplify things even further with classify-changes

calherries commented on src/metabase/models/dashboard.clj:
```diff
@@ -183,36 +184,42 @@
+                                           (if-let [new-tab-id (get old->new-tab-id (:dashboard_tab_id card))]
+                                             (assoc card :dashboard_tab_id new-tab-id)
+                                             card))))]
+    (revert-dashcards dashboard-id  serialized-cards))
```
```suggestion
    (revert-dashcards dashboard-id serialized-cards))
```

calherries reviewed (approved):
LGTM!

I think we should move classify-changes somewhere else though.

calherries commented on src/metabase/models/dashboard.clj:
```diff
@@ -258,6 +265,29 @@
+                 (set/subset? new-tab-ids prev-tab-ids)
+                 (> num-prev-tabs num-new-tabs))         (deferred-trun "removed a tab" "removed {0} tabs" num-tabs-diff)
+
+               (and (= num-prev-tabs num-new-tabs)
+                    (= prev-tab-ids new-tab-ids)
+                    (= (set (map :name prev-tabs))
+                       (set (map :name new-tabs))))      (deferred-tru "rearranged the tabs")
```
I think this is better expressed as "the set of tabs is the same excluding their positions"
```suggestion
               (= (set (map #(dissoc % :position) prev-tabs)
                  (set (map #(dissoc % :position) new-tabs))      (deferred-tru "rearranged the tabs")
```

qnkhuat replied:
Nice, I updated it here 6ba8e0b6ad

qnkhuat committed:
Move classify-changes to metabase.util.cljc
update diff change from Cal's suggestion

qnkhuat commented:
I moved `classify-changes` to metabase.util, I removed the annotation for types as I think the benefit of not overcomplicating the namespace + able to use it in js as well overweight the type annotation for this case.

qnkhuat committed:
Merge branch 'master' into ngoc-dashboard-tab-revision

qnkhuat committed:
remove unused ns

qnkhuat committed:
remove unused ns
